File: content/5-memory/2-history-compaction/aisdk/2-token_trim.ts
import { generateText, type ModelMessage } from "ai";
import { openai } from "@ai-sdk/openai";

const LLM_MODEL = "gpt-5.4";

// token-aware: only trim when approaching context limit
// estimate tokens from content length — AI SDK has no built-in token counter
const TOKEN_LIMIT = 100_000;
const CHARS_PER_TOKEN = 4; // rough estimate
const messages: ModelMessage[] = [];

function estimateTokens(msgs: ModelMessage[]): number {
  return msgs.reduce(
    (sum, m) => sum + String(m.content).length / CHARS_PER_TOKEN,
    0,
  );
}

async function chat(message: string): Promise<string> {
  messages.push({ role: "user", content: message });
  // short conversations stay intact, long ones get pruned
  let window: ModelMessage[] = messages;
  if (estimateTokens(messages) > TOKEN_LIMIT) {
    window = messages.slice(-10); // approaching limit — keep recent only
  }
  const result = await generateText({
    model: openai(LLM_MODEL),
    messages: window,
  });
  messages.push({ role: "assistant", content: result.text });
  return result.text;
}

console.log(await chat("What is the capital of France?"));
console.log(await chat("What is its population?"));

File: content/5-memory/2-history-compaction/pydantic/2-token_trim.py
from pydantic_ai import Agent, RunContext
from pydantic_ai.messages import ModelMessage

LLM_MODEL = "openai:gpt-5.4"

# token-aware: trim only when approaching the context window limit
# short conversations stay intact, long ones get pruned
# RunContext gives access to accumulated token usage

def token_trim(
    ctx: RunContext[None], messages: list[ModelMessage],
) -> list[ModelMessage]:
    if ctx.usage.total_tokens > 100_000:
        return messages[-10:]  # approaching limit — keep recent only
    return messages  # still fits — keep everything

agent = Agent(LLM_MODEL, history_processors=[token_trim])

history: list[ModelMessage] = []

result = agent.run_sync("What is the capital of France?", message_history=history)
history = result.new_messages()
print(result.output)

result = agent.run_sync("What is its population?", message_history=history)
history = result.new_messages()
print(result.output)
